### PR TITLE
Add JSON5 parsing and stringify support

### LINKED ISSUE #179: Add JSON5 support
## Summary

Add JSON5 parsing support as a separate, permissive structured-data format without changing the strict semantics of `JSON.parse(...)`.

## Motivation

JSON5 is widely used for hand-edited config files because it allows comments, trailing commas, single-quoted strings, and unquoted identifier keys. GocciaScript already supports JSON and YAML as configuration formats; JSON5 fills the gap for JSON-like configs that need to stay human-friendly.

## Current behavior

- `JSON.parse(...)` is strict, as it should be
- structured-data modules only recognize `.json`, `.yaml`, and `.yml`
- `--globals=file` accepts JSON or YAML by extension, but not JSON5
- there is no standalone JSON5 parser or `JSON5` builtin

## Expected behavior

- `JSON5.parse(text)` parses JSON5 into Goccia values
- `.json5` files can be imported through the structured-data module loader
- top-level object keys are exposed as named exports, matching `.json` module behavior
- `--globals=config.json5` works when the parsed root is an object

## Scope notes

- `JSON.parse(...)` must remain strict JSON; JSON5 support must be opt-in via a separate parser/builtin
- parse support is the priority; `JSON5.stringify(...)` is out of scope for this issue
- support the core JSON5 surface: comments, trailing commas, single-quoted strings, unquoted identifier keys, hex numbers, and `Infinity` / `NaN`
- do not broaden this issue into generic JSONC or editor-specific extensions unless they fall out naturally from the JSON5 grammar

## Proposed implementation plan

1. Add a standalone `Goccia.JSON5` unit with a `TGocciaJSON5Parser`, sharing helpers with `Goccia.JSON` where it is actually beneficial but keeping strict and permissive parsing behavior clearly separated.
2. Extend `Goccia.FileExtensions`, `IsStructuredDataExtension(...)`, and `TGocciaModuleLoader.LoadStructuredDataModule(...)` to recognize `.json5`.
3. Add a `TGocciaJSON5Builtin` with `JSON5.parse(text)` and register it behind a new `ggJSON5` global flag in the engine/runtime bootstrap.
4. Add globals-file support so `ScriptLoader` and embedding helpers can ingest `.json5` object files the same way they already ingest JSON and YAML.
5. Add end-to-end tests for JSON5 syntax support, strict/lenient boundary cases, module imports, globals injection, and docs updates.

### PR DESCRIPTION
**Pull request by frostney**
## Summary
- Added a new `JSON5` built-in with `JSON5.parse(...)` and `JSON5.stringify(...)` support.
- Extended structured-data handling so `.json5` files work for module imports, globals injection, and CLI/embedding documentation.
- Refactored the JSON parser into a capability-driven core so strict JSON and JSON5 share implementation while preserving their different syntax rules.
- Added an official JSON5 compatibility harness and end-to-end tests for parser behavior and stringify edge cases.

Fixes #179 

## Testing
- Not run: `./build.pas testrunner && ./build/TestRunner tests`
- Not run: `python3 scripts/run_json5_test_suite.py`
- Not run: `python3 scripts/run_json5_test_suite.py --harness=./build/GocciaJSON5Check`

## Summary by CodeRabbit

* **New Features**
  * JSON5 support: .json5 imports, structured globals by extension, CLI accepts JSON5 files, and runtime JSON5.parse()/JSON5.stringify() with replacer/space/quote options and JSON5 semantics.

* **Documentation**
  * Expanded README and docs with usage examples, parser/stringify options, design notes, and official JSON5 test‑suite guidance.

* **Tests**
  * Added comprehensive JSON5 parse/stringify/import tests and a deterministic compliance harness and extraction tooling.

* **CI/Tooling**
  * New JSON5 compliance runner and CI job producing per‑platform conformance reports.

### CODE REVIEW

File: docs/built-ins.md [hunk 1]
```diff
@@ -17,6 +17,7 @@ TGocciaGlobalBuiltin = (
   ggGlobalNumber,     // Number.parseInt, Number.isNaN, etc.
   ggPromise,          // Promise constructor, prototype, static methods, microtask queue
   ggJSON,             // JSON.parse, JSON.stringify
+  ggJSON5,            // JSON5.parse
   ggJSONL,            // JSONL.parse, JSONL.parseChunk
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/built-ins.md]: Fixed in 351a192. Updated the  built-ins documentation to list both  and .

File: units/Goccia.Builtins.JSON5.pas [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+          Obj.AssignProperty(PropKey, NewValue);
+      end;
+    end;
+  end;
+
+  Args := TGocciaArgumentsCollection.Create;
+  Args.Add(TGocciaStringLiteralValue.Create(AKey));
+  Args.Add(Value);
+  Result := InvokeCallable(AReviver, Args, AHolder);
+end;
```
[the remark this answers is not in the document]
Comment (frostney) [reply units/Goccia.Builtins.JSON5.pas]: Fixed in 351a192.  now frees its temporary  in a  block.

File: scripts/GocciaJSON5Check.dpr [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    on E: Exception do
+    begin
+      WriteLn(E.Message);
+      Halt(1);
+    end;
+  end;
+
+  Parser.Free;
+  Source.Free;
+  TGarbageCollector.Shutdown;
```
[the remark this answers is not in the document]
Comment (frostney) [reply scripts/GocciaJSON5Check.dpr]: Fixed in 27ee63c. The JSON5 compliance harness now uses normal control flow plus a real finally block for Parser.Free, Source.Free, and TGarbageCollector.Shutdown, so cleanup runs on both success and error paths.

File: scripts/run_json5_test_suite.py [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+          part for part in [process.stdout.strip(), process.stderr.strip()] if part
+        ),
+      }
+
+    report = json.loads(output_path.read_text(encoding="utf-8"))
+    report["ok"] = process.returncode == 0 and report.get("failed", 1) == 0
+    report["message"] = "\n".join(
+      part for part in [process.stdout.strip(), process.stderr.strip()] if part
+    )
+    return report
```
[the remark this answers is not in the document]
Comment (frostney) [reply scripts/run_json5_test_suite.py]: Fixed in c6b599c. evaluate_stringify_suite now takes the shared timeout value, passes it to subprocess.run, and returns a failed stringify report on timeout instead of hanging the compliance lane until the workflow timeout.

File: scripts/run_json5_test_suite.py [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+    "timeouts": 0,
+  }
+
+  with tempfile.TemporaryDirectory(prefix="json5-suite-cases.") as case_dir_name:
+    case_dir = Path(case_dir_name)
+
+    for index, case in enumerate(cases):
+      summary["total"] += 1
+      case_path = case_dir / f"case_{index}.json5"
+      case_path.write_text(case["source"], encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (frostney) [reply scripts/run_json5_test_suite.py]: Fixed in c6b599c. The JSON5 compliance runner now writes extracted case files with write_bytes(case["source"].encode("utf-8")) so Windows newline translation cannot mutate line-terminator-sensitive inputs before GocciaJSON5Check reads them.

File: docs/build-system.md [hunk 6]
```diff
@@ -260,13 +269,15 @@ Runs on the full platform matrix:
...
+**`artifacts`** (needs test + toml-compliance + json5-compliance + benchmark + examples, main only) — Uploads production binaries after all checks pass, deriving the executable names from the top-level `*.dpr` entrypoints.
 
-**`release`** (needs test + toml-compliance + benchmark + examples, tags only) — Downloads all platform build artifacts, stages only the shipped binaries derived from the top-level `*.dpr` entrypoints, bundles them with `tests/`, `benchmarks/`, and `examples/` into per-platform archives (`.tar.gz` for Linux/macOS, `.zip` for Windows), and creates a GitHub release using `softprops/action-gh-release`.
+**`release`** (needs test + toml-compliance + json5-compliance + benchmark + examples, tags only) — Downloads all platform build artifacts, stages only the shipped binaries derived from the top-level `*.dpr` entrypoints, bundles them with `tests/`, `benchmarks/`, and `examples/` into per-platform archives (`.tar.gz` for Linux/macOS, `.zip` for Windows), and creates a GitHub release using `softprops/action-gh-release`.
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/build-system.md]: Fixed in c6b599c. I rechecked the docs surface for --globals=file: README already mentioned JSON5, and docs/build-system.md was the stale occurrence. That page now includes a .json5 example and the format list now says JSON, JSON5, TOML, or YAML.
